### DISCUSSION TITLE
Help/question regarding static rendering

### DISCUSSION THREAD

**Discussion by levipadre** [amannn/next-intl#552] [Q&A]
Hi, thank you for creating next-intl. It is really really good.

I want to ask for some advice, or at least some help to explain something. maybe I misunderstood something basic.

On this blog page (https://next-intl-docs.vercel.app/blog/next-intl-3-0) it's been mentioned that:

> Static rendering of Server Components
With the newly introduced Server Components support comes a temporary workaround for static rendering.

> If you call APIs like useTranslations in a Server Component, by default, the hook will opt the page into dynamic rendering. This is a limitation that we aim to remove in the future, but as a stopgap solution, we've added the [unstable_setRequestLocale](https://next-intl-docs-git-feat-next-13-rsc-next-intl.vercel.app/docs/getting-started/app-router-server-components#static-rendering) API so that you can keep your pages fully static.

I thought I couldn't keep the pages fully static if I was using middleware. Or is middleware not required at all?
I have multiple domains. Only one page `src/app/[locale]/[[...slug]]/page.tsx` for handling every pages, where the data is coming from a restAPI. I need to recognise the host (at the moment I'm using `headers`), but if there is a better way to keep the pages static, it would be fantastic.

```
locales: ['en', 'fr', 'it', 'nl', 'pl', 'sv', 'cz', 'de', 'es'],
localeDetection: false,
localePrefix: 'never'
defaultLocale: 'en',
alternateLinks: false,
domains: [
    {
        domain: 'foo-domain.org.uk',
        defaultLocale: 'en',
        locales: ['en']
    },
    {
        domain: 'foo-domain.fr',
        defaultLocale: 'fr',
        locales: ['fr']
    },
    {
        domain: 'foo-domain.it',
        defaultLocale: 'it',
        locales: ['it']
    },
    {
        domain: 'foo-domain.nl',
        defaultLocale: 'nl',
        locales: ['nl']
    },
    {
        domain: 'foo-domain.pl',
        defaultLocale: 'pl',
        locales: ['pl']
    },
    {
        domain: 'foo-domain.se',
        defaultLocale: 'sv',
        locales: ['sv']
    },
    {
        domain: 'foo-domain.com',
        defaultLocale: 'en',
        locales: ['en']
    },
    {
        domain: 'foo-domain.cz',
        defaultLocale: 'cz',
        locales: ['cz']
    },
    {
        domain: 'foo-domain.de',
        defaultLocale: 'de',
        locales: ['de']
    },
    {
        domain: 'foo-domain.eu',
        defaultLocale: 'en',
        locales: ['en']
    },
    {
        domain: 'foo-domain.org',
        defaultLocale: 'en',
        locales: ['en']
    },
]
```

Thanks for any help.

**Comment by amannn** [comment]
> I thought I couldn't keep the pages fully static if I was using middleware. Or is middleware not required at all?

Next.js in fact makes a difference between [static rendering](https://nextjs.org/docs/app/building-your-application/rendering/server-components#static-rendering-default) of pages and [static exports](https://nextjs.org/docs/app/building-your-application/deploying/static-exports). The former can be combined with middleware, while the latter cannot.

> I need to recognise the host (at the moment I'm using headers), but if there is a better way to keep the pages static, it would be fantastic.

Not in an easy way at least. Maybe you're better off with dynamic rendering here. It's quite unfortunate that Next.js currently doesn't provide an option to add `cache-control` for dynamic routes, but possibly you could configure that at a proxy level as part of your deployment setup to speed up the pages.

**Reply by levipadre** [reply]
Hi @amannn. I revisited this question now that v3 is out. You mentioned something about static rendering in your announcement (https://next-intl-docs.vercel.app/blog/next-intl-3-0#static-rendering-of-server-components), but maybe I misunderstood.

Is it possible to get the locales without server-side rendering?

I tried this setup, but on a French site it still showed `en` locale.

- middleware:
 ```
 import createMiddleware from 'next-intl/middleware';
 
export default createMiddleware({
    locales: ['en', 'fr'],

    alternateLinks: false,
    
    defaultLocale: 'en',

    localeDetection: false,

    localePrefix: 'never',

    domains: [
        {
            domain: 'domain.org.uk',
            defaultLocale: 'en',
            locales: ['en']
        },
        {
            domain: 'domain.fr',
            defaultLocale: 'fr',
            locales: ['fr']
        }
    ],
});

export const config = {
    // Match only internationalized pathnames
    matcher: ['/((?!api|_next|_vercel|.*\\..*).*)',]
};
 ```
 
- layout (`src/app/[locale]/layout.tsx`):
```
import {unstable_setRequestLocale} from 'next-intl/server';
import { notFound } from 'next/navigation';

const locales = ['en', 'fr'];


export default async function RootLayout({
    children,
    params
}: { 
    children: React.ReactNode,
    params: {
        locale: string
    }
}) {
    if (!locales.includes(params.locale as any)) notFound();
    unstable_setRequestLocale(params.locale);

    return (
        <html lang={params.locale}> 
            <body>
                <h1>Layout {params.locale}</h1>
                {children}
            </body>
        </html>
    )
}
```

- page (`src/app/[locale]/page.tsx`):
```
import {unstable_setRequestLocale} from 'next-intl/server';

const locales = ['en', 'fr'];


export default async function Page({
    params
}: {
    params: {
        locale: string
    }
}) {
    unstable_setRequestLocale(params.locale);

    return (
        <main>
            <h2>Page {params.locale}</h2>
        </main>
    )
}
```

So when visit I any sites, it shows this:

<img width="267" alt="Screenshot 2023-11-16 at 10 43 52" src="https://github.com/amannn/next-intl/assets/357974/e04c9c62-4c86-422e-acf2-33200ac3453c">

**Reply by amannn** [reply]
> Is it possible to get the locales without server-side rendering?

Absolutely, yes!

I noticed you're using `localePrefix: 'never'`. Is the locale possibly resolved to `en` in your case? 

You can use the dev tools in Chrome to override the locale in your browser, maybe that helps to test all the locales your app supports.

<img width="420" alt="Screenshot 2023-11-16 at 12 13 29" src="https://github.com/amannn/next-intl/assets/4038316/0f67d6c2-0f72-4960-8ccc-b6ac9a4aace1">

**Reply by levipadre** [reply]
I need to recognise the site by the url and use that locale, just like I define in the middleware.
At the moment I can only get the site url (aka locale) by using headers, which makes the site dynamic rendered:
```
const headersList = headers();
const host = headersList.get('host') as string;
const locale = extractTLD(host);
```

I'll check localePrefix with different values, thanks.

**Reply by levipadre** [reply]
I tried both values for localePrefix but nothing changed, except for 'always', but in that case it added `/en` and `/fr`, which I don't need.
I simply need domain.fr and recognise the defined locale without server side rendering, just like I defined:
```
domains: [
    {
        domain: 'foo-domain.org.uk',
        defaultLocale: 'en',
        locales: ['en']
    },
    {
        domain: 'foo-domain.fr',
        defaultLocale: 'fr',
        locales: ['fr']
    },
    {
        domain: 'foo-domain.it',
        defaultLocale: 'it',
        locales: ['it']
    },
    {
        domain: 'foo-domain.nl',
        defaultLocale: 'nl',
        locales: ['nl']
    },
    {
        domain: 'foo-domain.pl',
        defaultLocale: 'pl',
        locales: ['pl']
    },
    {
        domain: 'foo-domain.se',
        defaultLocale: 'sv',
        locales: ['sv']
    },
    {
        domain: 'foo-domain.com',
        defaultLocale: 'en',
        locales: ['en']
    },
    {
        domain: 'foo-domain.cz',
        defaultLocale: 'cz',
        locales: ['cz']
    },
    {
        domain: 'foo-domain.de',
        defaultLocale: 'de',
        locales: ['de']
    },
    {
        domain: 'foo-domain.eu',
        defaultLocale: 'en',
        locales: ['en']
    },
    {
        domain: 'foo-domain.org',
        defaultLocale: 'en',
        locales: ['en']
    },
]
```

And as you can see, multiple domains have `en` locale too.

**Reply by amannn** [reply]
Ah, sorry—of course only the domain is relevant in your case.

Note this section from [the domain-based routing docs](https://next-intl-docs.vercel.app/docs/routing/middleware#domain-based-routing):

```
  // Used when no domain matches (e.g. on localhost)
  defaultLocale: 'en',
```

Therefore locally you should always see `en`.

If you need to read the host in render though, then your site will be dynamically rendered—yes.
